nethack/src/board.c

```diff
@@ -0,0 +1,226 @@
+    mdat->x1 = 0;
+    mdat->y1 = 0;
+    mdat->x2 = -1;
+    mdat->y2 = -1;
+
+    x1 = 0;
+    y1 = 0;
+    x2 = COLS;
+    y2 = LINES;
+
+    assign_menu_accelerators(mdat);
+    layout_menu(gw);
+
+    starty = (y2 - y1 - mdat->height) / 2 + y1;
+    startx = (x2 - x1 - mdat->width) / 2 + x1;
+
+    gw->win = newwin(mdat->height, mdat->width, starty, startx);
+    keypad(gw->win, TRUE);
+    wattron(gw->win, FRAME_ATTRS);
+    box(gw->win, 0, 0);
+    wattroff(gw->win, FRAME_ATTRS);
+    mdat->content =
+        derwin(gw->win, mdat->innerheight, mdat->innerwidth,
+               mdat->frameheight - 1, 2);
+    leaveok(gw->win, TRUE);
+    leaveok(mdat->content, TRUE);
+    wtimeout(gw->win, 500);
+}
+
+static void
+board_add_entry(struct nh_board_entry *entry, struct nh_menuitem **items,
+                int *size, int *icount, int maxwidth, int maxplayerwidth,
+                enum nh_menuitem_role txtrole)
+{
+    char fmtline[BUFSZ], line[BUFSZ], name[BUFSZ], hptxt[BUFSZ], entxt[BUFSZ];
+    struct tm tm;
+    time_t t, now;
+    char *inactive;
+
+    now = time(NULL);
+
+    strptime(entry->lastactive, "%Y-%m-%d %H:%M:%S.000000%z", &tm);
+    t = mktime(&tm);
+
+    if(difftime(now, t) > 30) { //Marks a player inactive after 30 seconds
+        inactive = "I";
+    } else {
+        inactive = "";
+    }
+
+    sprintf(name, "%s the %s %s %s %s", entry->name, entry->plalign,
+            entry->plgend, entry->plrace, entry->plrole);
+    sprintf(hptxt, "%d/%d", entry->hp, entry->hpmax);
+    sprintf(entxt, "%d/%d", entry->en, entry->enmax);
+    sprintf(fmtline, "%%1s  %%-%ds | %%-12s | %%-7s %%-7s | %%3d %%3d %%3d %%3d %%3d %%3d | %%5d | %%5d | %%5d",
+            maxplayerwidth);
+    sprintf(line, fmtline, inactive, name,
+            entry->leveldesc, hptxt, entxt, entry->wi, entry->in, entry->st,
+            entry->dx, entry->co, entry->ch, entry->moves, entry->depth,
+            entry->level);
+    add_menu_txt(*items, *size, *icount, line, txtrole);
+}
+
+static int
+get_max_playerdesc(struct nh_board_entry *entries, int listlen)
+{
+    int i, tmpsize, maxsize = 0;
+    struct nh_board_entry *tmpentry;
+    for(i = 0; i < listlen; i++) {
+        tmpentry = (entries + i);
+        tmpsize = strlen(tmpentry->name) //Length of the player's name
+                    + 5                  //Length of " the "
+                    + strlen(tmpentry->plalign) //Length of their alignment
+                    + 1                  //Space
+                    + strlen(tmpentry->plgend)
+                    + 1                  //Space
+                    + strlen(tmpentry->plrace)
+                    + 1                  //Space
+                    + strlen(tmpentry->plrole);
+        if(maxsize < tmpsize) {
+            maxsize = tmpsize;
+        }
+    }
+    return maxsize;
+}
+
+static void
+handle_key(int key)
+{
+    switch (key) {
+    case KEY_ESC:
+    case 'q':
+        done = TRUE;
+        draw_menu(gw);
```

Comment:
I'm not sure. You didn't end up actually answering my question. If it can fail, then you have to:

```
errno = 0;
draw_menu(gw);
if (errno != 0) it failed;
```

but if it can't fail, then you don't.